examples/trainer_controller/README.md

```diff
@@ -0,0 +1,121 @@
+
+- **Events**: Events supply the state and arguments required for the metric handlers to perform metric computation at the events they are registered for. The framework callback lists out all event handlers with prefix `"on_"` and loads then as event handlers. Every metric declares one or more events from this list of valid handlers. These computed metric variables are stored in a global state of the trainer controller and independently picked up the operations which could potentially be triggered on an entirely different set of events. This decouples the control loop for metrics and operations. I.e. the metric could be computed on event A, while operation could be triggered on event B. The controller rules which use the metric variables from the trainer controller state are evaluated and based on the outcomes specified actions are performed.
+![High-Level Design Diagram: Trainer Controller Framework](../../architecture_records/imgs/001-arch.png)
+
+## Usage and customization
+We added a new trainer callback (see [here](https://huggingface.co/docs/transformers/v4.37.2/en/main_classes/callback)) which accepts a `training control definition` YAML file containing:
+1. Rules to control training loop
+2. Trigger points that evaluate the above rules
+3. Control operation and action that needs to be performed if rule is evaluated to true.
+
+The trainer controller configuration is structured as shown below. There are list of metric definitions under `controller-metrics`, a list of operations and their actions under `operations` and a list of controllers, each of which define the rules, triggers and control operations.
+```yaml
+controller-metrics:
+  - name: <metric-name>
+    class: <metric-handler-class>
+    arguments:
+      <arg1>: <value1>
+      <arg2>: <value2>
+      ...
+operations:
+  - name: <operation-name>
+    class: <operation-handler-class>
+    arguments:
+      <arg1>: <value1>
+      <arg2>: <value2>
+      ...
+controllers:
+  - name: <controller-name>
+    triggers:
+      - <event-1>
+      ...
+    rule: <rule-string>
+    operations:
+      - <operation-action-1>
+      ...
+```
+The `controller-metrics` and `operations` are optional. We provide a set of built-in `controller-metrics` and `operations` which could be referred to without actually defining them as. For example, the below configuration defines a `controller-metric` called `loss` which refers to a built-in `Loss` controller-metric class with custom arguments (in this case, no arguments. If arguments are required, then they could be listed under a `arguments` section as shown above), but does not define any `operations`. It only refers to a built-in operation.
+```yaml
+controller-metrics:
+  - name: loss
```

Comment:
how can I know if it is validation loss or train loss?

Reply:
This is explained in the **Metrics** bullet point in this [section](https://github.com/foundation-model-stack/fms-hf-tuning/blob/1ca546d5050e1f6d4ba52441343fb04dc2423b89/examples/trainercontroller_configs/Readme.md#usage).